test/functional/page_objects/visual_builder_page.ts

```diff
@@ -69,80 +83,177 @@ export function VisualBuilderPageProvider({ getService, getPageObjects }) {
+      }
+      const variables: any[] = await testTableVariables.findAllByCssSelector(variablesSelector);
+
+      const variablesKeyValueSelectorMap = await Promise.all(
+        variables.map(async (variable: any) => {
+          const subVars = await variable.findAllByCssSelector('td');
+          const selector = await subVars[0].findByTagName('a');
+          const key = await selector.getVisibleText();
+          const value = await subVars[1].getVisibleText();
+          log.debug(`markdown table variables table is: ${key} ${value}`);
+          return { key, value, selector };
+        })
+      );
+      return variablesKeyValueSelectorMap;
+    }
+
+    /**
+     * return variable table message, if `table` is empty it will be fail
+     *
+     * **Note:** if `table` have variables, use `getMarkdownTableVariables` method instead
+     * @see {@link VisualBuilderPage#getMarkdownTableVariables}
+     * @returns
+     * @memberof VisualBuilderPage
+     */
+    public async getMarkdownTableNoVariables() {
+      return await testSubjects.getVisibleText('tvbMarkdownEditor__noVariables');
+    }
+
+    /**
+     * get all sub-tabs count for `time series`, `metric`, `top n`, `gauge`, `markdown` or `table` tab.
+     *
+     * @returns {Promise<any[]>}
+     * @memberof VisualBuilderPage
+     */
+    public async getSubTabs() {
+      return await find.allByCssSelector('[data-test-subj$="-subtab"]');
+    }
+
+    /**
+     * switch markdown sub-tab for visualization
+     *
+     * @param {'data' | 'options'| 'markdown'} subTab
+     * @memberof VisualBuilderPage
+     */
+    public async markdownSwitchSubTab(subTab: 'data' | 'options' | 'markdown') {
+      const element = await testSubjects.find(`${subTab}-subtab`);
+      await element.click();
+    }
+
+    /**
+     * get all sub-tabs count for `time series`, `metric`, `top n`, `gauge`, `markdown` or `table` tab.
+     *
+     * @returns {Promise<number>}
+     * @memberof VisualBuilderPage
+     */
+    public async getSubTabsCount() {
+      const rootSelector = await find.byCssSelector('.euiTabs.euiTabs--small');
+      const tabsSelector = 'button[role="tab"]';
+      const elements = await rootSelector.findAllByCssSelector(tabsSelector);
+
+      return elements.length;
     }
 
-    async clickMarkdownData() {
-      await testSubjects.click('markdownDataBtn');
+    /**
+     * switch sub-tab for visualization
+     *
+     * @param {'Data' | 'Panel options'| 'Annotations' | 'Markdown' | 'Columns'} subTab
+     * @memberof VisualBuilderPage
+     */
+    public async switchSubTab(
+      subTab: 'Data' | 'Panel options' | 'Annotations' | 'Markdown' | 'Columns'
+    ) {
+      const rootSelector = await find.byCssSelector('.euiTabs.euiTabs--small');
+      const tabsSelector = `button[role="tab"] span`;
+
+      const elements = await rootSelector.findAllByCssSelector(tabsSelector);
+
+      elements.forEach((tab: any) => {
+        if (tab.getVisibleText() === subTab) {
```

Comment:
async operation, missing `await`.
Double check how to deal with `forEach+async`: https://codeburst.io/javascript-async-await-with-foreach-b6ba62bbf404